### ISSUE TITLE
Make use of `vcpkg` friendly for new developers

### ISSUE THREAD

**Issue by DanRStevens** [lairworks/nas2d-core#373]
We should find a way to make use of `vcpkg` more friendly for new developers. That might involve better error messages, or increased automation.

Some important tasks:
- Installing `vcpkg`
- Updating `vcpkg`
- Running the `InstallVcpkgDeps.bat` file to install dependencies
- Updating dependencies

We should try to automate it as much as possible. Where automation is not likely to work, we should provide easy to read and understand error messages. We might even add web links for places they need to go to download `vcpkg`.

**Comment by cugone** [comment]
I found an automated script that clones a git repo if it doesn't exist, otherwise pulls it:

https://gist.github.com/nicferrier/2277987

There are some interesting additions to it in the later comments.

**Comment by DanRStevens** [comment]
I like the last comment with the example use of the `-C` option to run a `git` command without having to change the current working directory.

**Comment by cugone** [comment]
Except `/dev/null` isn't a thing on windows. :P

**Comment by DanRStevens** [comment]
Not anymore than the shell syntax. :wink:

----

I did some experimenting with Visual Studio project file features, which might be relevant here.

In #365 I made brief mention of the [`Exec` task](https://docs.microsoft.com/en-us/visualstudio/msbuild/exec-task?view=vs-2019).

Example used to capture the `vcpkg` install path:
```
  <Target Name="Find vcpkg" BeforeTargets="PrepareForBuild">
    <Exec Command="where vcpkg" ConsoleToMsBuild="true" >
      <Output TaskParameter="ConsoleOutput" PropertyName="VcpkgExePath" />
    </Exec>
  </Target>
```

Different sources seem to swap out `PropertyName` with `ItemName`, depending on use.

We could potentially use the `Exec` task to automatically install dependencies. We can likely even include a `Condition` which checks if dependency folders exist. I haven't experimented with the condition for that, though running the batch file might look something like:
```
  <Target Name="Install vcpkg dependencies" BeforeTargets="PrepareForBuild">
    <Exec Command="call InstallVcpkgDeps.bat" />
  </Target>
```

You can set variables (conditionally) in the `Globals` section. I expect you can use your own custom sections too:
```
  <PropertyGroup Label="Globals">
    ...
    <VcpkgTriplet Condition="'$(Platform)'=='Win32'">x86-windows</VcpkgTriplet>
    <VcpkgTriplet Condition="'$(Platform)'=='x64'">x64-windows</VcpkgTriplet>
    <VcpkgInclude>$(VcpkgExePath)\..\installed\$(VcpkgTriplet)\include</VcpkgInclude>
  </PropertyGroup>
```

However, I found the variables set in the `Globals` section were set before the `Exec` task was run (placed later in the file), so I wasn't able to use the output to form the variables. There might be a way to do it, either moving sections around, or somehow specifying run order dependencies.

**Comment by cugone** [comment]
Since you're better at manual project file manipulation since, you know, you have to, I'll let you handle this. I can check the results locally to see if VS refuses to load the project or any other major errors.

**Comment by cugone** [comment]
I got vcpkg working local to and exclusively with NAS2D via `vcpkg integrate project` and copying a generated NuGet package install command into the Package Manager Console. That said, it modifies the `.vcxproj` file with potentially personal or sensitive information in the form of a local absolute path to the generated local NuGet package as well as adding a `packages` folder to the project root.

Keeping these changes is required for it to work, but pushing them to the repo will break everyone else's project as it is locally exclusive to whomever generated the NuGet package (as well as a security/privacy concern).

Is there a way to ignore a file/folder, make changes, then unignore them? Determining that will allow me to write up some instructions in the ReadMe.

**Comment by DanRStevens** [comment]
Yeah, I remember reading somewhere that integrating `vcpkg` with a project through the NuGet package feature wasn't recommended for shared projects with multiple developers. I'd never really explored why though.

Ignoring the `packages` folder would be easy enough with `.gitignore`. Though handling the project file modifications would be more difficult. Perhaps you could conditionally load project settings from another file, and have that other file listed in `.gitignore`. I'm pretty sure the project file syntax allows for conditionally loading settings from another file. Pretty sure that support is already used with some of the system files for the Visual Studio project settings, and to handle plugins, or user specific settings.

Checking the project file, I see external settings being conditionally imported with:
```xml
<Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
```

Could do similar with a file loaded from the project folder. I assume paths would by default be relative to the project folder.

**Comment by cugone** [comment]
Stashing the changes seems to be a workaround. I'll see what happens when I unstash when the repo has newer commits on top of it.

**Comment by cugone** [comment]
> 
> Stashing the changes seems to be a workaround. I'll see what happens when I unstash when the repo has newer commits on top of it.

This is a viable workaround (using a GUI version of Git will make it *mush* easier):

0. Clone vcpkg.
1. Run `./bootstrap-vcpkg.bat`.
2. Install vcpkg via `vcpkg integrate project`.
3. Run the generated install command in the Package Manager Console.
4. Building the NAS2D solution will also install the required dependencies.
5. Stash (with an appropriate name) the NuGet-changes to `.vcxproj`, `.vcxproj.filters`, `./packages.config`, and `./packages/...`
6. Do any other work/commits.
7. Pop/Unstash the NuGet-changes.
8. Compile.
9. Re-stash the NuGet-changes.

**Comment by DanRStevens** [comment]
Related Issue:
- https://github.com/OutpostUniverse/OPHD/issues/1374

Documentation for `vcpkg` Manifests
https://vcpkg.readthedocs.io/en/latest/users/manifests/

**Comment by DanRStevens** [comment]
There, and all we had to do was wait 3 years for `vcpkg` to be upgraded to support project specific manifests.

**Comment by cugone** [comment]
Safe to say this closes #952 ?
